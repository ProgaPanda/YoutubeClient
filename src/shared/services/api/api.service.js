import axios from 'axios';

const baseURL = 'https://www.googleapis.com/youtube/v3/';

const api = axios.create({
  baseURL,
});

export default {
  /**
   * Returns matching items for the search parameter from YT API
   * @param {string} searchParam
   */
  search: (searchParam) =>
    api
      .get(
        `search?type=video&part=snippet&order=viewCount&q=${searchParam}+tutorial&maxResults=3&key=AIzaSyBT4Pe1v-xPdTqYBc8FJy5THAtUg90C4-0`,
      )
      .then((d) => console.log(d)),
};
