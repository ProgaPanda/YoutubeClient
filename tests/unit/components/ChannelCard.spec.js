import { shallowMount } from '@vue/test-utils';
import ChannelCard from '@/components/ChannelCard.component.vue';
import api from '@/shared/services/api/api.service';

const mockedApiResponse = {
  data: {
    items: [
      {
        kind: 'youtube#channel',
        id: 'UC29ju8bIPH5as8OGnQzwJyA',
        snippet: {
          title: 'Traversy Media',
          description:
            'Traversy Media features the best online web development and programming tutorials',
          thumbnails: {
            medium: {
              url: 'https://yt3.ggpht.com/thumb',
            },
          },
        },
        statistics: {
          subscriberCount: 1080000,
          videoCount: 784,
        },
      },
    ],
  },
};

// Mocking API layer
jest.mock('@/shared/services/api/api.service', () => ({
  getChannelDetails: jest.fn(() => Promise.resolve(mockedApiResponse)),
}));

describe('ChannelCard', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(ChannelCard, {
      id: 'id',
      title: 'title',
      channel: 'String',
      thumbnail: 'url',
      description: 'description',
      publishedAt: '3 years ago',
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('calls api module with getChannelDetails after mount', () => {
    const CHANNEL_ID = 'MOCK_ID';
    shallowMount(ChannelCard, {
      propsData: {
        id: CHANNEL_ID,
      },
    });

    expect(api.getChannelDetails).toBeCalledWith(CHANNEL_ID);
  });

  it('updates videos count and subscribers count after getChannelDetails request', () => {
    const wrapper = shallowMount(ChannelCard, {
      propsData: {
        id: 'id',
        title: 'title',
        channel: 'String',
        thumbnail: 'url',
        description: 'description',
        publishedAt: '3 years ago',
      },
    });

    expect(wrapper.vm.subscribeCount).toEqual(0);
    expect(wrapper.vm.videosCount).toEqual(0);

    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.subscribeCount).toEqual(1080000);
      expect(wrapper.vm.videosCount).toEqual(784);
    });
  });

  it('does not update videos count and subscribers count after getChannelDetails request: fail scenario', () => {
    api.getChannelDetails.mockImplementationOnce(() => {
      return Promise.reject(new Error());
    });

    const wrapper = shallowMount(ChannelCard);
    expect(wrapper.vm.subscribeCount).toEqual(0);
    expect(wrapper.vm.videosCount).toEqual(0);

    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.subscribeCount).toEqual(0);
      expect(wrapper.vm.videosCount).toEqual(0);
    });
  });
});
