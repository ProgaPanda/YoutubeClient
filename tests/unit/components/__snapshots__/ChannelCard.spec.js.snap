// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelCard renders correctly 1`] = `
<div
  class="channel-card"
>
  <div
    class="channel-card__thumbnail"
  >
    <img
      alt="video-thumbnail"
    />
  </div>
   
  <div
    class="channel-card__details"
  >
    <h4
      class="channel-card__details__title"
    >
      
    </h4>
     
    <p
      class="channel-card__details__sub-count"
    >
      
      0 subscribers â€¢ 
      <span>
        0 videos
      </span>
    </p>
     
    <p
      class="channel-card__details__description"
    >
      
    </p>
  </div>
</div>
`;
